File: bench/scala/src/main/scala/ming/SchemeReader.scala
package ming

private[ming] object SchemeReader:
  import SchemeInterpreter.Expr

  def readAll(input: String): List[Expr] =
    Reader(input).readAll()

  final private class Reader(input: String):
    private var index = 0
    private var line  = 1
    private var col   = 1

    private def currentPos: SourcePos =
      SourcePos(line, col)

    def readAll(): List[Expr] =
      val expressions = List.newBuilder[Expr]
      skipTrivia()
      while index < input.length do
        expressions += readExpr()
        skipTrivia()
      expressions.result()

    private def readExpr(): Expr =
      skipTrivia()
      if index >= input.length then throw EvalError.at(currentPos, "unexpected end of input")

      val pos = currentPos
      input.charAt(index) match
        case '\'' =>
          advance()
          Expr.ListExpr(List(Expr.Symbol("quote", pos), readExpr()), pos)
        case '(' =>
          advance()
          readList(pos)
        case ')' =>
          throw EvalError.at(pos, "unexpected )")
        case '"' =>
          readString(pos)
        case '#' =>
          readHashLiteral(pos)
        case _ =>
          readAtom(pos)

    private def readList(startPos: SourcePos): Expr =
      val items = List.newBuilder[Expr]
      skipTrivia()
      while index < input.length && input.charAt(index) != ')' do
        items += readExpr()
        skipTrivia()

      if index >= input.length then throw EvalError.at(startPos, "unterminated list")

      advance()
      Expr.ListExpr(items.result(), startPos)

    private def readString(startPos: SourcePos): Expr =
      advance()
      val builder = new StringBuilder
      var closed  = false

      while index < input.length && !closed do
        val ch = advance()
        ch match
          case '"' =>
            closed = true
          case '\\' =>
            if index >= input.length then throw EvalError.at(startPos, "unterminated string escape")
            val escaped = advance()
            builder.append(
              escaped match
                case '"'   => '"'
                case '\\'  => '\\'
                case 'n'   => '\n'
                case 't'   => '\t'
                case other => other
            )
          case other =>
            builder.append(other)

      if !closed then throw EvalError.at(startPos, "unterminated string")

      Expr.StringLit(builder.result(), startPos)

    private def readHashLiteral(startPos: SourcePos): Expr =
      if startsWithToken("#t") then
        advance()
        advance()
        Expr.Bool(true, startPos)
      else if startsWithToken("#f") then
        advance()
        advance()
        Expr.Bool(false, startPos)
      else if input.startsWith("#\\", index) then readCharacter(startPos)
      else throw EvalError.at(startPos, "invalid boolean literal")

    private def readCharacter(startPos: SourcePos): Expr =
      advance()
      advance()

      val start = index
      while index < input.length && !isDelimiter(input.charAt(index)) do advance()

      val token = input.substring(start, index)
      val value = token match
        case "space"            => ' '
        case "newline"          => '\n'
        case s if s.length == 1 => s.charAt(0)
        case _                  => throw EvalError.at(startPos, "invalid character literal")

      Expr.Character(value, startPos)

    private def readAtom(startPos: SourcePos): Expr =
      val start = index
      while index < input.length && !isDelimiter(input.charAt(index)) do advance()

      val token = input.substring(start, index)
      if isIntegerToken(token) then Expr.Number(BigInt(token), startPos)
      else Expr.Symbol(token, startPos)

    private def skipTrivia(): Unit =
      var keepSkipping = true
      while keepSkipping do
        while index < input.length && input.charAt(index).isWhitespace do advance()

        if index < input.length && input.charAt(index) == ';' then
          while index < input.length && input.charAt(index) != '\n' do advance()
        else keepSkipping = false

    private def startsWithToken(token: String): Boolean =
      input.startsWith(token, index) && {
        val boundary = index + token.length
        boundary >= input.length || isDelimiter(input.charAt(boundary))
      }

    private def isDelimiter(ch: Char): Boolean =
      ch.isWhitespace || ch == '\'' || ch == '(' || ch == ')' || ch == '"' || ch == ';'

    private def isIntegerToken(token: String): Boolean =
      token.nonEmpty &&
        (token.forall(_.isDigit) ||
          (token.head == '-' && token.length > 1 && token.tail.forall(_.isDigit)))

    private def advance(): Char =
      val ch = input.charAt(index)
      index += 1
      if ch == '\n' then
        line += 1
        col = 1
      else col += 1
      ch

  private object Reader:
    def apply(input: String): Reader = new Reader(input)

File: bench/java/src/main/java/ming/SchemeReader.java
package ming;

import java.util.ArrayList;
import java.util.List;

/**
 * Tokenizer and parser for Scheme source text.
 * Converts raw input strings into S-expression trees.
 */
class SchemeReader {

    record Token(String value, int line, int col) {}

    record Located(Object expr, int line, int col) {}

    static Object unwrap(Object o) {
        return o instanceof Located loc ? loc.expr() : o;
    }

    @SuppressWarnings("unchecked")
    static Object deepUnwrap(Object o) {
        o = unwrap(o);
        if (o instanceof List<?> list) {
            List<Object> result = new ArrayList<>(list.size());
            for (Object elem : list) result.add(deepUnwrap(elem));
            return result;
        }
        return o;
    }

    List<Token> tokenize(String input) throws EvalError {
        List<Token> tokens = new ArrayList<>();
        int i = 0;
        int len = input.length();
        int line = 1;
        int col = 1;
        while (i < len) {
            char c = input.charAt(i);
            if (c == '\n') {
                line++;
                col = 1;
                i++;
            } else if (Character.isWhitespace(c)) {
                col++;
                i++;
            } else if (c == ';') {
                while (i < len && input.charAt(i) != '\n') {
                    i++;
                    col++;
                }
            } else if (c == '\'') {
                tokens.add(new Token("'", line, col));
                i++;
                col++;
            } else if (c == '`') {
                tokens.add(new Token("`", line, col));
                i++;
                col++;
            } else if (c == ',') {
                if (i + 1 < len && input.charAt(i + 1) == '@') {
                    tokens.add(new Token(",@", line, col));
                    i += 2;
                    col += 2;
                } else {
                    tokens.add(new Token(",", line, col));
                    i++;
                    col++;
                }
            } else if (c == '(') {
                tokens.add(new Token("(", line, col));
                i++;
                col++;
            } else if (c == ')') {
                tokens.add(new Token(")", line, col));
                i++;
                col++;
            } else if (c == '"') {
                int startLine = line;
                int startCol = col;
                StringBuilder sb = new StringBuilder();
                sb.append('"');
                i++;
                col++;
                while (i < len && input.charAt(i) != '"') {
                    if (input.charAt(i) == '\\') {
                        sb.append(input.charAt(i));
                        i++;
                        col++;
                        if (i < len) {
                            sb.append(input.charAt(i));
                            i++;
                            col++;
                        }
                    } else {
                        if (input.charAt(i) == '\n') {
                            line++;
                            col = 1;
                        } else {
                            col++;
                        }
                        sb.append(input.charAt(i));
                        i++;
                    }
                }
                if (i < len) {
                    sb.append('"');
                    i++;
                    col++;
                }
                tokens.add(new Token(sb.toString(), startLine, startCol));
            } else if (c == '#') {
                int startCol = col;
                if (i + 1 < len) {
                    char next = input.charAt(i + 1);
                    if (next == 't') {
                        tokens.add(new Token("#t", line, startCol));
                        i += 2;
                        col += 2;
                    } else if (next == 'f') {
                        tokens.add(new Token("#f", line, startCol));
                        i += 2;
                        col += 2;
                    } else if (next == '\\') {
                        if (i + 2 < len) {
                            int charStart = i + 2;
                            int charEnd = charStart;
                            while (charEnd < len && !Character.isWhitespace(input.charAt(charEnd))
                                    && input.charAt(charEnd) != ')' && input.charAt(charEnd) != '('
                                    && input.charAt(charEnd) != '"' && input.charAt(charEnd) != ';') {
                                charEnd++;
                            }
                            String charName = input.substring(charStart, charEnd);
                            String tok = "#\\" + charName;
                            tokens.add(new Token(tok, line, startCol));
                            int tokLen = tok.length();
                            i += tokLen;
                            col += tokLen;
                        } else {
                            tokens.add(new Token("#\\", line, startCol));
                            i += 2;
                            col += 2;
                        }
                    } else if (next == '(') {
                        tokens.add(new Token("#(", line, startCol));
                        i += 2;
                        col += 2;
                    } else if (next == '\'') {
                        tokens.add(new Token("#'", line, startCol));
                        i += 2;
                        col += 2;
                    } else {
                        String sym = readSymbol(input, i);
                        tokens.add(new Token(sym, line, startCol));
                        i += sym.length();
                        col += sym.length();
                    }
                } else {
                    tokens.add(new Token("#", line, startCol));
                    i++;
                    col++;
                }
            } else {
                int startCol = col;
                String sym = readSymbol(input, i);
                tokens.add(new Token(sym, line, startCol));
                i += sym.length();
                col += sym.length();
            }
        }
        return tokens;
    }

    private String readSymbol(String input, int start) {
        int i = start;
        int len = input.length();
        while (i < len) {
            char c = input.charAt(i);
            if (Character.isWhitespace(c) || c == '(' || c == ')' || c == '"' || c == ';' || c == '\'') {
                break;
            }
            i++;
        }
        return input.substring(start, i);
    }

    Object parse(List<Token> tokens, int[] pos) throws EvalError {
        if (pos[0] >= tokens.size()) {
            throw new EvalError("unexpected end of input");
        }
        Token token = tokens.get(pos[0]);
        pos[0]++;

        if (token.value().equals("'")) {
            Object quoted = parse(tokens, pos);
            List<Object> quoteExpr = new ArrayList<>();
            quoteExpr.add(new SchemeSymbol("quote"));
            quoteExpr.add(quoted);
            return new Located(quoteExpr, token.line(), token.col());
        }

        if (token.value().equals("`")) {
            Object body = parse(tokens, pos);
            List<Object> expr = new ArrayList<>();
            expr.add(new SchemeSymbol("quasiquote"));
            expr.add(body);
            return new Located(expr, token.line(), token.col());
        }

        if (token.value().equals(",")) {
            Object body = parse(tokens, pos);
            List<Object> expr = new ArrayList<>();
            expr.add(new SchemeSymbol("unquote"));
            expr.add(body);
            return new Located(expr, token.line(), token.col());
        }

        if (token.value().equals(",@")) {
            Object body = parse(tokens, pos);
            List<Object> expr = new ArrayList<>();
            expr.add(new SchemeSymbol("unquote-splicing"));
            expr.add(body);
            return new Located(expr, token.line(), token.col());
        }

        if (token.value().equals("#'")) {
            Object syntaxed = parse(tokens, pos);
            List<Object> syntaxExpr = new ArrayList<>();
            syntaxExpr.add(new SchemeSymbol("syntax"));
            syntaxExpr.add(syntaxed);
            return new Located(syntaxExpr, token.line(), token.col());
        }

        if (token.value().equals("#(")) {
            List<Object> elems = new ArrayList<>();
            while (pos[0] < tokens.size() && !tokens.get(pos[0]).value().equals(")")) {
                elems.add(parse(tokens, pos));
            }
            if (pos[0] >= tokens.size()) throw new EvalError("missing closing parenthesis");
            pos[0]++;
            List<Object> vecForm = new ArrayList<>();
            vecForm.add(new SchemeSymbol("vector"));
            vecForm.addAll(elems);
            return new Located(vecForm, token.line(), token.col());
        }

        if (token.value().equals("(")) {
            List<Object> list = new ArrayList<>();
            while (pos[0] < tokens.size() && !tokens.get(pos[0]).value().equals(")")) {
                list.add(parse(tokens, pos));
            }
            if (pos[0] >= tokens.size()) {
                throw new EvalError("missing closing parenthesis");
            }
            pos[0]++;
            return new Located(list, token.line(), token.col());
        } else if (token.value().equals(")")) {
            throw new EvalError("unexpected )");
        } else {
            return new Located(parseAtom(token.value()), token.line(), token.col());
        }
    }

    private Object parseAtom(String token) {
        if (token.equals("#t")) {
            return Boolean.TRUE;
        }
        if (token.equals("#f")) {
            return Boolean.FALSE;
        }
        if (token.startsWith("#\\")) {
            String charName = token.substring(2);
            if (charName.equals("space")) return new SchemeChar(' ');
            if (charName.equals("newline")) return new SchemeChar('\n');
            if (charName.equals("tab")) return new SchemeChar('\t');
            if (charName.length() == 1) return new SchemeChar(charName.charAt(0));
            return new SchemeChar(charName.charAt(0));
        }
        if (token.startsWith("\"") && token.endsWith("\"")) {
            return token;
        }
        try {
            return Long.parseLong(token);
        } catch (NumberFormatException e) {}
        int slash = token.indexOf('/');
        if (slash > 0 && slash < token.length() - 1) {
            try {
                long num = Long.parseLong(token.substring(0, slash));
                long den = Long.parseLong(token.substring(slash + 1));
                return SchemeRational.make(num, den);
            } catch (NumberFormatException e) {}
        }
        try {
            return Double.parseDouble(token);
        } catch (NumberFormatException e) {}
        return new SchemeSymbol(token);
    }
}
